securecookie.go

```diff
@@ -15,19 +15,102 @@ import (
 	"encoding/base64"
 	"encoding/gob"
 	"encoding/json"
-	"errors"
 	"fmt"
 	"hash"
 	"io"
 	"strconv"
+	"strings"
 	"time"
 )
 
-var (
-	errNoCodecs      = errors.New("securecookie: no codecs provided")
-	errHashKeyNotSet = errors.New("securecookie: hash key is not set")
+// Error is the interface of all errors returned by functions in this library.
+type Error interface {
+	error
+
+	// IsUsage returns true for errors indicating the client code probably
+	// uses this library incorrectly.  For example, the client may have
+	// failed to provide a valid hash key, or may have failed to configure
+	// the Serializer adequately for encoding value.
+	IsUsage() bool
+
+	// IsDecode returns true for errors indicating that a cookie could not
+	// be decoded and validated.  Since cookies are usually untrusted
+	// user-provided input, errors of this type should be expected.
+	// Usually, the proper action is simply to reject the request.
+	IsDecode() bool
+
+	// IsInternal returns true for unexpected errors occurring in the
+	// securecookie implementation.
+	IsInternal() bool
+
+	// Cause, if it returns a non-nil value, indicates that this error was
+	// propagated from some underlying library.  If this method returns nil,
+	// this error was raised directly by this library.
+	//
+	// Cause is provided principally for debugging/logging purposes; it is
+	// rare that application logic should perform meaningfully different
+	// logic based on Cause.  See, for example, the caveats described on
+	// (MultiError).Cause().
+	Cause() error
+}
+
+// errorType is a bitmask giving the error type(s) of an errorImpl value.
+type errorType int
+
+const (
+	usageError = errorType(1 << iota)
+	decodeError
+	internalError
+)
+
+type errorImpl struct {
+	typ   errorType
+	msg   string
+	cause error
+}
+
+func (e errorImpl) IsUsage() bool    { return (e.typ & usageError) != 0 }
+func (e errorImpl) IsDecode() bool   { return (e.typ & decodeError) != 0 }
+func (e errorImpl) IsInternal() bool { return (e.typ & internalError) != 0 }
 
-	ErrMacInvalid = errors.New("securecookie: the value is not valid")
+func (e errorImpl) Cause() error { return e.cause }
+
+func (e errorImpl) Error() string {
+	parts := []string{"securecookie: "}
+	if e.msg == "" {
+		parts = append(parts, "error")
+	} else {
+		parts = append(parts, e.msg)
+	}
+	if c := e.Cause(); c != nil {
+		parts = append(parts, " - caused by: ", c.Error())
```

Comment:
I like this—the `Cause()` method is similar to Rust's error trait. Although it's rare to need to know the root cause (as you have pointed out earlier in the file) the ability to inspect a little deeper can be useful for debugging.